src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2576,6 +2576,65 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
+        /// It's also possible that some newer ALTER_METADATA entries are present in source_queue list,
+        /// and source replica are executing such entry right not (or had executed recently).
+        /// More than that, /metadata_version update is not atomic with /columns and /metadata update...
+
+        /// Fortunately, ALTER_METADATA seems to be idempotent,
+        /// and older entries of such type can be replaced with newer entries.
+        /// Let's try to get consistent values of source replica's /columns and /metadata
+        /// and prepend dummy ALTER_METADATA to our replication queue.
+        /// It should not break anything if source_queue already contains ALTER_METADATA entry
+        /// with greater or equal metadata_version, but it will update our metadata
+        /// if all such entries were cleaned up from the log and source_queue.
+
+        LOG_WARNING(log, "Metadata version ({}) on replica is not up to date with metadata ({}) on source replica {}",
+                    metadata_version, source_metadata_version, source_replica);
+
+        String source_metadata;
+        String source_columns;
+        while (true)
+        {
+            Coordination::Stat metadata_stat;
+            Coordination::Stat columns_stat;
+            source_metadata = zookeeper->get(source_path + "/metadata", &metadata_stat);
+            source_columns = zookeeper->get(source_path + "/columns", &columns_stat);
+
+            Coordination::Requests ops;
+            Coordination::Responses responses;
+            ops.emplace_back(zkutil::makeCheckRequest(source_path + "/metadata", metadata_stat.version));
+            ops.emplace_back(zkutil::makeCheckRequest(source_path + "/columns", columns_stat.version));
+
+            Coordination::Error code = zookeeper->tryMulti(ops, responses);
+            if (code == Coordination::Error::ZOK)
+                break;
+            else if (code == Coordination::Error::ZBADVERSION)
+                LOG_WARNING(log, "Metadata of replica {} was changed", source_path);
+            else
+                zkutil::KeeperMultiException::check(code, ops, responses);
+        }
+
+        ReplicatedMergeTreeLogEntryData dummy_alter;
+        dummy_alter.type = LogEntry::ALTER_METADATA;
+        dummy_alter.source_replica = source_replica;
+        dummy_alter.metadata_str = source_metadata;
+        dummy_alter.columns_str = source_columns;
+        dummy_alter.alter_version = source_metadata_version;
+        dummy_alter.create_time = time(nullptr);
+        /// We don't really know if it has mutation or not, but it doesn't matter.
+        dummy_alter.have_mutation = false;
+
+        zookeeper->create(replica_path + "/queue/queue-", dummy_alter.toString(), zkutil::CreateMode::PersistentSequential);
+
+        /// TODO Do we also need to do something with mutation_pointer? Seems like we don't, but it's not clear from the code.
```

Comment:
Definitely yes, we shouldn't do anything here, but our mutation pointer must be equal with the source replica's pointer?

Reply:
No, it may be less than mutation pointer of the source replica